router/dataplane.go

```diff
@@ -209,41 +211,57 @@ func (d *DataPlane) SetIA(ia addr.IA) (e error) {
 
 // SetKey sets the key used for MAC verification. The key provided here should
 // already be derived as in scrypto.HFMacFactory.
+// Verified locally in 29m45s with --disableMoreCompleteExhale --parallelizeBranches
+// Marked as trusted for now because of timeouts
 // @ trusted
-// @ requires  false
+// @ requires  acc(&d.key,        1/2)
+// @ requires  acc(d.key,         1/2)
 // @ requires  acc(&d.running,    1/2) && !d.running
 // @ requires  acc(&d.macFactory, 1/2) && d.macFactory == nil
-// @ requires  len(key) == 0
+// @ requires  len(key) > 0
 // @ requires  slices.AbsSlice_Bytes(key, 0, len(key))
 // @ preserves d.mtx.LockP()
 // @ preserves d.mtx.LockInv() == MutexInvariant!<d!>;
 // @ ensures   acc(&d.running,    1/2) && !d.running
-// @ ensures   acc(&d.macFactory, 1/2)
-func (d *DataPlane) SetKey(key []byte) error {
-	//@ share key
+// @ ensures   res == nil ==> d.MacFactoryOperational()
+func (d *DataPlane) SetKey(key []byte) (res error) {
+	// @ share key
 	d.mtx.Lock()
 	defer d.mtx.Unlock()
-	//@ unfold MutexInvariant!<d!>()
-	//@ defer fold MutexInvariant!<d!>()
+	// @ unfold MutexInvariant!<d!>()
+	// @ defer fold MutexInvariant!<d!>()
 	if d.running {
+		// @ def.Unreachable()
 		return modifyExisting
 	}
 	if len(key) == 0 {
+		// @ def.Unreachable()
 		return emptyValue
 	}
 	if d.macFactory != nil {
+		// @ def.Unreachable()
 		return alreadySet
 	}
 	// First check for MAC creation errors.
 	if _, err := scrypto.InitMac(key); err != nil {
 		return err
 	}
-	// (TODO)
-	// (VerifiedSCION) Gobra crashes due to the closure literal.
-	d.macFactory = func /*@ f @*/ () hash.Hash {
-		mac, _ := scrypto.InitMac(key)
-		return mac
-	}
+	// @ d.key = &key
+	verScionTemp :=
```

Comment:
do we still need this temporary variable? (It is also fine if we keep it though, let's not wait so much time to re-verify because of a variable renaming)

Reply:
Yes, it is still needed